### PR TITLE
ddl: fix alter table charset bug and compatibility

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
1. Change table charset should change all column charset too.

In MySQL
```SQL
mysql > create table t1 (a varchar(10) character set ascii);
Query OK, 0 rows affected
Time: 0.021s
mysql > show  create table t1
+-------+----------------------------------------------------+
| Table | Create Table                                       |
+-------+----------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                |
|       |   `a` varchar(10) CHARACTER SET ascii DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=latin1             |
+-------+----------------------------------------------------+
1 row in set
Time: 0.007s
mysql > alter table t1 convert to charset utf8mb4;
Query OK, 0 rows affected
Time: 0.028s
mysql > show  create table t1
+-------+-----------------------------------------+
| Table | Create Table                            |
+-------+-----------------------------------------+
| t1    | CREATE TABLE `t1` (                     |
|       |   `a` varchar(10) DEFAULT NULL          |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 |
+-------+-----------------------------------------+
1 row in set
```

In TiDB 
```SQL
tidb > create table t1 (a varchar(10) character set ascii) charset=utf8
Query OK, 0 rows affected
Time: 0.007s
tidb > show create table t1
+-------+----------------------------------------------------------------------+
| Table | Create Table                                                         |
+-------+----------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                  |
|       |   `a` varchar(10) CHARACTER SET ascii COLLATE ascii_bin DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin                |
+-------+----------------------------------------------------------------------+
1 row in set
Time: 0.006s
tidb > alter table t1 convert to charset utf8mb4;
Query OK, 0 rows affected
Time: 0.007s
tidb > show create table t1
+-------+----------------------------------------------------------------------+
| Table | Create Table                                                         |
+-------+----------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                  |
|       |   `a` varchar(10) CHARACTER SET ascii COLLATE ascii_bin DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin          |
+-------+----------------------------------------------------------------------+
1 row in set
Time: 0.006s
```

2. The  table charset  maybe "" cause create in the old TiDB version(such as 2.0.8), then execute alter table charset will return error, But it should execute successfully.

3. Some behaver is not compatible with MySQL.

In MySQL
```sql
create table t (a varchar(10) character set utf8) charset=utf8;
mysql > alter table t charset latin1 charset utf8 charset utf8mb4 collate utf8_bin;
(1302, u"Conflicting declarations: 'CHARACTER SET latin1' and 'CHARACTER SET utf8'")
mysql > alter table t charset utf8mb4 collate "" collate utf8mb4_bin;
(1273, u"Unknown collation: ''")
mysql > alter table t charset utf8mb4 collate "";
(1273, u"Unknown collation: ''")
mysql > alter table t collate "";
(1273, u"Unknown collation: ''")
```

In TiDB v2.0.8
```sql
create table t1 (a varchar(10));
```

```shell
$ curl "http://$IP:10080/schema/test/t1"
{
    ...
    "charset": "",    #table charset.
    "collate": "",
    "cols": [
        {
            ...
            "id": 1,
            "name": {
                "L": "a",
                "O": "a"
            },
            ...
            "type": {
                "Charset": "utf8",  #column charset.
                "Collate": "utf8_bin",
                 ...
```

In TiDB master, 
```sql
tidb > alter table t1 convert to charset utf8mb4;
(1105, u'unsupported modify charset from  to utf8mb4')
```

### What is changed and how it works?
Change AlterTableCharsetAndCollate func.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch  2.1

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -2470,6 +2459,49 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
...
+			return true, nil
+		}
+	}
+
+	if len(origCharset) == 0 {
+		// The table charset may be "", if the table is create in old TiDB version.
+		// This DDL will update the table charset to default charset.
+		// Use default charset or database charset?
+		// show create table is uses the default charset.
+		origCharset, origCollate = charset.GetDefaultCharsetAndCollate()
```
Comment (crazycs520) [inline ddl/ddl_api.go]: Maybe both ok, because It will convert to `toCharset `  if below check pass.
@winkyao PTAL

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2472,6 +2461,49 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
...
+			if col.Charset == charset.CharsetBin {
+				continue
+			}
+			if col.Charset == toCharset && col.Collate == toCollate {
+				continue
+			}
+			doNothing = false
+		}
+		if doNothing {
+			return true, nil
```
Comment (xiekeyi98) [inline ddl/ddl_api.go]:
```suggestion
			return doNothing, err
```
I think this change will more like Golang style.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -2472,6 +2461,49 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 	return errors.Trace(err)
 }
 
+func checkAlterTableCharset(tblInfo *model.TableInfo, toCharset, toCollate string) (bool, error) {
+	var err error
+	origCharset := tblInfo.Charset
+	origCollate := tblInfo.Collate
+	if origCharset == toCharset && origCollate == toCollate {
+		// nothing to do.
+		var doNothing = true
```
Comment (xiekeyi98) [inline ddl/ddl_api.go]:
```suggestion
		doNoting := true
```
Or
```suggestion
		var doNothing bool = true
```

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2472,6 +2461,49 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
...
+	if len(origCharset) == 0 {
+		// The table charset may be "", if the table is create in old TiDB version.
+		// This DDL will update the table charset to default charset.
+		// Use default charset or database charset?
+		// show create table is uses the default charset.
+		origCharset, origCollate = charset.GetDefaultCharsetAndCollate()
+	}
+
+	if err = modifiableCharsetAndCollation(toCharset, toCollate, origCharset, origCollate); err != nil {
+		return false, errors.Trace(err)
```
Comment (xiekeyi98) [inline ddl/ddl_api.go]: Do we need `errors.Trace` now?

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -2472,6 +2461,49 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
...
+	}
+
+	for _, col := range tblInfo.Columns {
+		if col.Tp == mysql.TypeVarchar {
+			if err = IsTooBigFieldLength(col.Flen, col.Name.O, toCharset); err != nil {
+				return false, errors.Trace(err)
+			}
+		}
+	}
+	return false, nil
```
Comment (xiekeyi98) [inline ddl/ddl_api.go]: ditto.
I think we should better return the variable, rather constant.

File: ddl/ddl_api.go (hunk 4 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: No need any more. already fix this. PTAL

File: ddl/ddl_api.go (hunk 5 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: done.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -2472,6 +2461,48 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 	return errors.Trace(err)
 }
 
+func checkAlterTableCharset(tblInfo *model.TableInfo, toCharset, toCollate string) (doNothing bool, err error) {
```
Comment (shenli) [inline ddl/ddl_api.go]: Please add comments for this function.

Comment (crazycs520) [reply ddl/ddl_api.go]: done.

File: ddl/ddl_api.go (hunk 1 above)
Comment (winkyao) [reply ddl/ddl_api.go]: use database charset is better.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -224,18 +224,27 @@ func ResolveCharsetCollation(tblCharset, dbCharset string) (string, string, erro
 	return charset, collate, nil
 }
 
+func typesNeedCharset(tp byte) bool {
```
Comment (zimulala) [inline ddl/ddl_api.go]: I think this name is a bit ambiguous. Other types also need charset.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -2536,12 +2545,9 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
...
-
-	origCharset := tb.Meta().Charset
-	origCollate := tb.Meta().Collate
 	if toCharset == "" {
 		// charset does not change.
-		toCharset = origCharset
+		toCharset = tb.Meta().Charset
 	}
 
 	if toCollate == "" {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply ddl/ddl_api.go]: What if toCharset != nil and toCharset != table.Charset?

File: ddl/ddl_api.go (hunk 7 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: `typesNeedSetCharset`?

File: ddl/ddl_api.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl_api.go]: It seems that if `toCollate` is nil, MySQL used the default collate.
```
mysql> show create table t;
+-------+--------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` varchar(10) COLLATE utf8_bin DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+--------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> alter table t convert to charset utf8mb4;
Query OK, 0 rows affected (0.03 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> show create table t;
+-------+-------------------------------------------------------------------------------------------+
| Table | Create Table                                                                              |
+-------+-------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 |
+-------+-------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql>  alter table t convert to charset utf8 collate utf8_bin;
Query OK, 0 rows affected (0.04 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> alter table t convert to charset utf8;
Query OK, 0 rows affected (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> show create table t;
+-------+----------------------------------------------------------------------------------------+
| Table | Create Table                                                                           |
+-------+----------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl_api.go]: So we needn't change it.

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -2536,12 +2545,9 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 	if err != nil {
 		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
 	}
-
-	origCharset := tb.Meta().Charset
-	origCollate := tb.Meta().Collate
 	if toCharset == "" {
 		// charset does not change.
-		toCharset = origCharset
+		toCharset = tb.Meta().Charset
```
Comment (zimulala) [inline ddl/ddl_api.go]: Do we need to put an error here?
```
mysql> show create table t;
+-------+----------------------------------------------------------------------------------------+
| Table | Create Table                                                                           |
+-------+----------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> alter table t convert to collate utf8_bin;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'collate utf8_bin' at line 1
```

File: ddl/ddl_api.go (hunk 7 above)
Comment (bb7133) [reply ddl/ddl_api.go]: IMHO this name is OK, there is `has_charset()` accessor in MySQL's field definition: https://github.com/mysql/mysql-server/blob/ca94b993454c86be248fbe180db94647488114e9/sql/field.h#L1565

But I think it's better to move this function to `parser`(later)

File: ddl/db_integration_test.go [hunk 10]
```diff
@@ -587,6 +589,83 @@ func (s *testIntegrationSuite) TestChangingTableCharset(c *C) {
 
 	rs, err = tk.Exec("alter table t charset utf8mb4 collate utf8mb4_bin")
 	c.Assert(err, NotNil)
+
+	// Test change column charset when change table charset.
```
Comment (bb7133) [inline ddl/db_integration_test.go]: change -> changing

File: ddl/ddl_api.go (hunk 9 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: `alter table t collate "";`
Done. PTAL.

File: ddl/db_integration_test.go (hunk 10 above)
Comment (crazycs520) [reply ddl/db_integration_test.go]: done.

File: ddl/ddl_api.go (hunk 7 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: Both ok for me. currently is `typesNeedSetCharset`. So which one is better?

Comment (zimulala) [reply ddl/ddl_api.go]: I don't think the name is good, but I don't have any better suggestions, so I can accept both.

**Comment by bb7133** [comment]
LGTM

File: ddl/ddl_api.go (hunk 9 above)
Comment (zimulala) [reply ddl/ddl_api.go]: It seems that the commit of https://github.com/pingcap/tidb/pull/9790/commits/e87ad6b5dcc01301bba45ed993ca83973a6702b1 is disappeared.

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -1590,11 +1599,21 @@ func getCharsetAndCollateInTableOption(startIdx int, options []*ast.TableOption)
...
+		charset = charsets[0]
 	}
-
 	if len(collates) != 0 {
+		for i := range collates {
+			if collates[i] == "" {
+				return "", "", ErrUnknownCollation.GenWithStackByArgs("")
+			}
+		}
 		collate = collates[len(collates)-1]
```
Comment (zimulala) [inline ddl/ddl_api.go]: It seems that we need to check all collates.
```
mysql> alter table t charset utf8 collate utf8_bin collate utf8_bin collate utf8mb4_bin;
ERROR 1253 (42000): COLLATION 'utf8mb4_bin' is not valid for CHARACTER SET 'utf8'
mysql> alter table t charset utf8 collate utf8_bin collate utf8mb4_bin collate utf8_bin;
ERROR 1253 (42000): COLLATION 'utf8mb4_bin' is not valid for CHARACTER SET 'utf8'
```

**Pull request merged**